### ISSUE TITLE
Add documentation on how to use alive-progress outside of a context manager situation

### ISSUE THREAD

**Issue by zackees** [rsalmei/alive-progress#246]
Hi, I have a multi threaded application. It's very unclear how to use this solution outside of a with-statement.

Is this use case straightforward? Can it be added to the readme?

Thank you

**Comment by rsalmei** [comment]
Hi, zackees,

Thanks for the suggestion, but I won't ever do that. I strongly recommend against running alive-progress outside of a `with` context. I've made it intentionally like this so I could correctly initialize the background thread and install the system print and logging hooks on start, as well as stop the thread and uninstall the hooks on shutdown. Any use besides this is a hack, that's why I won't ever document that.

I've once seen someone actually needing such a hack, it was a very unusual scenario where he received asynchronous calls from Azure informing a client about some background processing. You can find that closed issue here if you really want it, but I don't see the relation between a multi-threaded application and this hack.
Anyway, multi-threaded applications should generally only run alive-progress on the main thread, and collect results of completed tasks or status updates via some `gather` or `map` pool methods, or a `Queue`.

I'm going to close this, but feel free to continue discussing any more doubts here.
You're welcome!

**Comment by zackees** [comment]
My scenario is that I wanted to use this for s3 uploading in parts, which takes a callback.

I don't see why whatever parts you need to be done can't be protected under a lock. But maybe there is something I'm missing.

**Comment by rsalmei** [comment]
Humm, in this case of asynchronous callbacks I'd recommend queues.
Start a thread that receives the callbacks with a queue to be your producer, and monitor status updates on the main thread via the same queue, the consumer.

**Comment by zackees** [comment]
"Rewrite your codebase because this lib forces you to use a context" is not a good answer.

A progress bar should come at the end of the dev cycle, when we want to make things pretty. Everything works and has unit tests. I'm not going to re-write it for the gui especially when less pretty but totally acceptable solutions exist.

**Comment by rsalmei** [comment]
I do trust them, that's why I helped once to do this exact thing in another issue, like I told you, you would just have to go and copy its code. I just said I won't ever document it, because it is a hack nevertheless.

And I suggested another approach for you, which I consider far better and does not require extensive changes at all, nor any async code for that matter. It was a much more beautiful encapsulation with separation of concerns, but I think it is a moot point now since you've gone with another lib.
